### PR TITLE
Refactor bindings parallel flow tests

### PR DESCRIPTION
**Pull request by alonre24**
**Describe the changes in the pull request**

 - Refactor Python bindings - includes avoiding using Python code while allowing generic.
- Add bulk insert/search functions that run in parallel for HNSW -  add vector /  KNN search / range search.
- Add parallel tests for insert/search various scenarios (single/multi index + knn/range/batched searches) - measure recall and throughput.

### CODE REVIEW

File: tests/flow/test_hnswlib.py [hunk 1]
```diff
@@ -518,11 +450,481 @@ def test_multi_range_query():
...
+    bf_params.metric = metric
+    bf_params.type = data_type
+    bf_params.initialCapacity = num_elements
+    bf_params.blockSize = num_elements
+    bf_params.multi = True
+
+    bf_index = BFIndex(bf_params)
+
+    np.random.seed(47)
+    data = np.random.random((num_labels, per_label, dim))
```
Comment (meiravgri) [inline tests/flow/test_hnswlib.py:698]: is it OK to not specify the datatype? will its size be 8 bytes?

Comment (meiravgri) [inline tests/flow/test_hnswlib.py:698]:
```suggestion
    data = np.float64(np.random.random((num_elements, per_label, dim)))
```
?

File: tests/flow/test_hnswlib.py [hunk 2]
```diff
@@ -518,11 +450,481 @@ def test_multi_range_query():
...
+    assert parallel_multi_index.index_size() == 0
+
+    def insert_vectors():
+        parallel_multi_index.add_vector_parallel(data, labels, num_threads=int(n_threads/2))
+
+    res_labels_g = np.zeros((num_queries, dim))
+
+    def run_queries():
+        nonlocal res_labels_g
+        res_labels_g, _ = parallel_multi_index.knn_parallel(query_data, k, num_threads=int(n_threads/2))
```
Comment (meiravgri) [inline tests/flow/test_hnswlib.py:777]: why not to use searchKnnParallel?

File: tests/flow/test_hnswlib.py [hunk 3]
```diff
@@ -518,11 +450,481 @@ def test_multi_range_query():
...
+    for i in range(0, num_queries, chunk_size):
+        total_correct_cur_chunk = 0
+        for j in range(i, i+chunk_size):
+            total_correct_cur_chunk += len(set(total_res_bf[j]).intersection(set(res_labels_g[j])))
+        if i == chunk_size:  # first iteration, there is no previous chunk
+            total_correct_prev_chunk = total_correct_cur_chunk
+        else:
+            assert total_correct_cur_chunk >= total_correct_prev_chunk
+            total_correct_prev_chunk = total_correct_cur_chunk
+        print(f"Recall for chunk {int(i/chunk_size)+1}/{int(num_queries/chunk_size)} of queries is:"
```
Comment (meiravgri) [inline tests/flow/test_hnswlib.py:804]:
```suggestion
        print(f"Recall for queries' chunk {int(i/chunk_size)+1}/{int(num_queries/chunk_size)} is:"
```
Took me a while to understand this sentence 😅

File: tests/flow/test_hnswlib.py [hunk 4]
```diff
@@ -518,11 +450,481 @@ def test_multi_range_query():
...
+    total_correct_prev_chunk = 0
+    for i in range(0, num_queries, chunk_size):
+        total_correct_cur_chunk = 0
+        for j in range(i, i+chunk_size):
+            total_correct_cur_chunk += len(total_res_bf[j].intersection(total_results_parallel[j]))
+        if i == chunk_size:  # first iteration, there is no previous chunk
+            total_correct_prev_chunk = total_correct_cur_chunk
+        else:
+            assert total_correct_cur_chunk >= total_correct_prev_chunk
+            total_correct_prev_chunk = total_correct_cur_chunk
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply tests/flow/test_hnswlib.py:928]: also, in what scenario we will get total_correct_cur_chunk < total_correct_prev_chunk
can len be negative?

**Review by meiravgri — COMMENTED** [review]
Good and robust job :) !
few comments - 
I think that the getdatabylabel should be templated because of the unsupervised memory allocation which is very dangerous. We anyway check the type in the bindings before calling it.
Also, I think the test can be somehow split where possible, and replace the performance asserts with noticeable logs so we can continue running even if one test fails (like in the unit tests)

Added a few more comments along the way :), see below

File: src/python_bindings/bindings.cpp [hunk 5]
```diff
@@ -63,86 +74,181 @@ class PyBatchIterator {
...
+        py::array query(input);
         return PyBatchIterator(
-            index, VecSimBatchIterator_New(index.get(), input_to_blob(input), query_params));
+            index, VecSimBatchIterator_New(index.get(), (const char *)query.data(0), query_params));
     }
 
-    virtual ~PyVecSimIndex() {} // Delete function was given to the shared pointer object
+    py::object getVector(labelType label) {
+        std::vector<const char *> res;
+        reinterpret_cast<VecSimIndexInterface *>(this->index.get())->getDataByLabel(label, res);
```
[the remark this answers is not in the document]
Comment (alonre24) [reply src/python_bindings/bindings.cpp:178]: I agree about moving res initialization to `rawVectorsAsNumpy`. I also agree that `getDataByLabel` can be removed from `VecSimIndexInterface`, but I think that this test function should be available for every index in general.

File: src/python_bindings/bindings.cpp [hunk 6]
```diff
@@ -63,86 +74,181 @@ class PyBatchIterator {
...
+        if (index->info().algo == VecSimAlgo_BF) {
+            size_t dim = index->info().bfInfo.dim;
+            if (index->info().bfInfo.type == VecSimType_FLOAT32) {
+                return rawVectorsAsNumpy<float>(res, dim);
+            } else if (index->info().bfInfo.type == VecSimType_FLOAT64) {
+                return rawVectorsAsNumpy<double>(res, dim);
+            } else {
+                throw std::runtime_error("Invalid vector data type");
```
[the remark this answers is not in the document]
Comment (alonre24) [reply src/python_bindings/bindings.cpp:186]: `rawVectorsAsNumpy`, but in the case of an error there's no release (as there should also be no allocation, as `getDataByLabel` is undefined...) Anyway, moving the call into `rawVectorsAsNumpy` should solve this issue.

File: src/python_bindings/bindings.cpp [hunk 7]
```diff
@@ -63,86 +74,181 @@ class PyBatchIterator {
...
-        VecSimQueryResult_List res =
-            VecSimIndex_TopKQuery(index.get(), input_to_blob(input), k, query_params, BY_SCORE);
-        if (VecSimQueryResult_Len(res) != k) {
-            throw std::runtime_error("Cannot return the results in a contiguous 2D array. Probably "
-                                     "ef or M is too small");
+    py::object knn(const py::object &input, size_t k, VecSimQueryParams *query_params) {
+        py::array query(input);
+        VecSimQueryResult_List res;
+        {
+            py::gil_scoped_release py_gil;
```
[the remark this answers is not in the document]
Comment (alonre24) [reply src/python_bindings/bindings.cpp:152]: To enable parallel `knn` from different Python threads. The creation of `py::gil_scoped_release` object releases the GIL and reacquires it when it goes out of scope (then we need to hold the GIL for replying)

File: src/python_bindings/bindings.cpp [hunk 8]
```diff
@@ -63,86 +74,181 @@ class PyBatchIterator {
...
-    py::object knn(py::object input, size_t k, VecSimQueryParams *query_params) {
-        VecSimQueryResult_List res =
-            VecSimIndex_TopKQuery(index.get(), input_to_blob(input), k, query_params, BY_SCORE);
-        if (VecSimQueryResult_Len(res) != k) {
-            throw std::runtime_error("Cannot return the results in a contiguous 2D array. Probably "
-                                     "ef or M is too small");
+    py::object knn(const py::object &input, size_t k, VecSimQueryParams *query_params) {
+        py::array query(input);
+        VecSimQueryResult_List res;
+        {
```
[the remark this answers is not in the document]
Comment (alonre24) [reply src/python_bindings/bindings.cpp:151]: see previous comment

File: src/python_bindings/bindings.cpp [hunk 9]
```diff
@@ -63,86 +74,181 @@ class PyBatchIterator {
...
-    const char *input_to_blob(py::object input) {
-        tmp_bytearray = create_bytearray(input);
-        return PyByteArray_AS_STRING(tmp_bytearray.ptr());
+    template <typename searchParam> // size_t/double for KNN/range queries.
+    using QueryFunc =
+        std::function<VecSimQueryResult_List(const char *, searchParam, VecSimQueryParams *)>;
+
+    template <typename searchParam> // size_t/double for KNN / range queries.
+    void runParallelQueries(const py::array &queries, size_t n_queries, searchParam param,
+                            VecSimQueryParams *query_params, int n_threads,
```
[the remark this answers is not in the document]
Comment (alonre24) [reply src/python_bindings/bindings.cpp:213]: Since we use -1 as default for `n_threads`, which stands for the number of threads as the number of CPUs in the machine.

File: src/python_bindings/bindings.cpp [hunk 10]
```diff
@@ -155,11 +261,115 @@ class PyHNSWLibIndex : public PyVecSimIndex {
...
+        size_t n_queries = queries.shape(0);
+        std::function<VecSimQueryResult_List(const char *, double, VecSimQueryParams *)>
+            searchRangeWrapper([this](const char *query_, double radius_,
+                                      VecSimQueryParams *query_params_) -> VecSimQueryResult_List {
+                return this->searchRangeInternal(query_, radius_, query_params_);
+            });
+        VecSimQueryResult_List results[n_queries];
+        runParallelQueries<double>(queries, n_queries, radius, query_params, n_threads,
+                                   searchRangeWrapper, results);
+        size_t max_results_num = 1;
```
[the remark this answers is not in the document]
Comment (alonre24) [reply src/python_bindings/bindings.cpp:297]: I did it to avoid the case of creating a matrix with `0` as one of its axes in `wrap_results`, which may complicate things in the corner cases where all queries returned 0 results. It doesn't really matter, we use padding in the redundant places anyway...

File: tests/flow/test_hnswlib.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (alonre24) [reply tests/flow/test_hnswlib.py:928]: It's possible since we set `total_correct_cur_chunk` to 0 in every iteration of the outer loop

File: tests/flow/test_hnswlib.py (hunk 1 above)
Comment (alonre24) [reply tests/flow/test_hnswlib.py:698]: yes, the default is float64

File: tests/flow/test_hnswlib.py (hunk 2 above)
Comment (alonre24) [reply tests/flow/test_hnswlib.py:777]: what do you mean? `searchKnnParallel` is an internal function in the Python bindings. `knn_parallel` is the method name that we expose in the Python module...

File: tests/flow/test_hnswlib.py (hunk 1 above)
Comment (alonre24) [reply tests/flow/test_hnswlib.py:698]: the default for numpy data type is float64

File: tests/flow/test_hnswlib.py [hunk 11]
```diff
@@ -518,11 +450,481 @@ def test_multi_range_query():
...
+        total_correct_parallel += len(total_res_bf[i].intersection(set(res_labels[i])))
+
+    print(f"Running parallel search, got {total_correct_parallel / (k * num_queries)} recall on {num_queries} queries,"
+          f" average query time is {total_search_time_parallel / num_queries} seconds")
+    print(f"Got {total_search_time / total_search_time_parallel} times improvement in runtime using {n_threads} threads\n")
+
+    # Validate that the recall of the parallel search recall is the same as the sequential search recall.
+    assert total_correct_parallel == total_correct
+    # Validate that the parallel run managed to achieve at least (n_threads - 1) times improvement in total runtime.
+    assert total_search_time / total_search_time_parallel > expected_speedup
```
[the remark this answers is not in the document]
Comment (alonre24) [reply tests/flow/test_hnswlib.py:542]: I know that some of these are flaky, but I want to have an assert so we'll see it fails if we change something (no one will check the logs unless he sees a red X...). I tried to set a threshold that is not too strict, but still meaningful.

**Review by meiravgri — APPROVED** [review]

File: tests/flow/test_hnswlib.py [hunk 12]
```diff
@@ -518,11 +450,522 @@ def test_multi_range_query():
...
+        batch_iterator_ = index.create_batch_iterator(query_)
+        res_labels_ = set()
+        for _ in range(n_batches):
+            res_labels_ = res_labels_.union(set(batch_iterator_.get_next_results(batch_size, BY_SCORE)[0][0]))
+        total_results_parallel[query_ind] = res_labels_
+
+    start = time.time()
+    with ThreadPoolExecutor(max_workers=n_threads) as executor:
+        futures = [executor.submit(run_batched_search, q, i) for i, q in enumerate(query_data)]
+        done, not_done = wait(futures, return_when=concurrent.futures.ALL_COMPLETED)
```
Comment (DvirDukhan) [inline tests/flow/test_hnswlib.py:916]: I think you can create a function that gets a collection of batch iterators and runs them in parallel, same as with KNN and range.

File: src/VecSim/algorithms/hnsw/hnsw.h [hunk 13]
```diff
@@ -235,6 +235,19 @@ class HNSWIndex : public VecSimIndexAbstract<DistType>,
...
+     * @brief Used for testing - store vector(s) data associated with a given label. This function
+     * copies the vector(s)' data buffer(s) and place it in the output vector
+     *
+     * @param label
+     * @param vectors_output empty vector to be modified, should store the blob(s) associated with
+     * the label.
+     */
+    virtual void getDataByLabel(labelType label,
+                                std::vector<std::vector<DataType>> &vectors_output) const = 0;
+#endif
```
[the remark this answers is not in the document]
Comment (alonre24) [reply src/VecSim/algorithms/hnsw/hnsw.h:249]: But we don't have `DataType` template in vecsim interface...

File: tests/flow/test_hnswlib.py [hunk 14]
```diff
@@ -518,11 +450,522 @@ def test_multi_range_query():
...
+        bf_index.add_vector(vector, i)
+
+    query_data = np.float32(np.random.random((num_queries, dim)))
+
+    # Collect the ground truth results
+    total_res_bf = []  # save the ground truth
+    for i, query in enumerate(query_data):
+        total_res_bf.append(set(bf_index.knn_query(query, k)[0][0]))
+
+    # Run search over non-parallel index as the baseline
```
[the remark this answers is not in the document]
Comment (alonre24) [reply tests/flow/test_hnswlib.py:584]: The comment is confusing. I run parallel search over an index which is "not parallel" (that is, was built by inserting vectors sequentially)

File: tests/flow/test_hnswlib.py [hunk 15]
```diff
@@ -518,11 +450,522 @@ def test_multi_range_query():
...
+    total_correct_parallel = 0
+    for i in range(num_queries):
+        total_correct_parallel += len(total_res_bf[i].intersection(set(res_labels[i])))
+    print(f"Running parallel search on index that was created using parallel insert, got "
+          f"{total_correct_parallel / (k * num_queries)} recall on {num_queries} queries, average query time is"
+          f" {total_search_time_parallel / num_queries} seconds")
+    assert total_correct_parallel >= total_correct * 0.95
+
+    # Insert vectors to the index and search in parallel.
+    parallel_index = create_hnsw_index(dim, num_elements, VecSimMetric_Cosine, VecSimType_FLOAT32, ef_runtime=200)
```
[the remark this answers is not in the document]
Comment (alonre24) [reply tests/flow/test_hnswlib.py:609]: what?

File: tests/flow/test_hnswlib.py [hunk 16]
```diff
@@ -518,11 +450,522 @@ def test_multi_range_query():
...
+        query_results_set.discard(PADDING_LABEL)  # remove the irrelevant padding values
+        assert query_results_set.issubset(set(total_res_bf[i]))
+        overall_intersection_rate_parallel += len(query_results_set) / total_res_bf[i].size \
+            if total_res_bf[i].size > 0 else 1
+    print(f"Running the same {num_queries} queries in parallel, average query time is"
+          f" {total_range_query_parallel_time/num_queries}, and intersection rate is: "
+          f"{overall_intersection_rate_parallel/num_queries}")
+    assert overall_intersection_rate_parallel == overall_intersection_rate
+    print(f"Got improvement of {total_search_time/total_range_query_parallel_time} times using {n_threads} threads\n")
+    assert total_search_time/total_range_query_parallel_time >= expected_speedup
```
[the remark this answers is not in the document]
Comment (alonre24) [reply tests/flow/test_hnswlib.py:713]: why so?

File: tests/flow/test_hnswlib.py (hunk 12 above)
Comment (alonre24) [reply tests/flow/test_hnswlib.py:916]: I know, but I preferred doing it that way for the batch iterator. I'll explain when we'll talk

**Review by DvirDukhan — APPROVED** [review]
